neqo-transport/src/send_stream.rs

```diff
@@ -1063,61 +1115,290 @@ impl ::std::fmt::Display for SendStream {
+                // already set as fair.
+
+                // This normally is only called when a new stream is created.  If
+                // so, because of how we allocate StreamIds, it should always have
+                // the largest value.  This means we can just append it to the
+                // regular vector.  However, if we were ever to change this
+                // invariant, things would break subtly.
+
+                // To be safe we can try to insert at the end and if not
+                // fall back to binary-search insertion
+                if matches!(self.regular.stream_ids().last(), Some(last) if stream_id > *last) {
+                    self.regular.stream_ids().push(stream_id);
+                } else {
+                    self.regular.insert_streamid(&stream_id);
+                }
+            } else if old_fair && !make_fair {
+                // remove from the OrderGroup
+                let group = if let Some(sendorder) = stream.sendorder {
+                    self.sendordered.get_mut(&sendorder).unwrap()
+                } else {
+                    &mut self.regular
+                };
+                group.remove_streamid(&stream_id);
+            }
+            Ok(())
+        } else {
+            Err(Error::InvalidStreamId)
+        }
     }
 
     pub fn acked(&mut self, token: &SendStreamRecoveryToken) {
-        if let Some(ss) = self.0.get_mut(&token.id) {
+        if let Some(ss) = self.map.get_mut(&token.id) {
             ss.mark_as_acked(token.offset, token.length, token.fin);
         }
     }
 
     pub fn reset_acked(&mut self, id: StreamId) {
-        if let Some(ss) = self.0.get_mut(&id) {
+        if let Some(ss) = self.map.get_mut(&id) {
             ss.reset_acked()
         }
     }
 
     pub fn lost(&mut self, token: &SendStreamRecoveryToken) {
-        if let Some(ss) = self.0.get_mut(&token.id) {
+        if let Some(ss) = self.map.get_mut(&token.id) {
             ss.mark_as_lost(token.offset, token.length, token.fin);
         }
     }
 
     pub fn reset_lost(&mut self, stream_id: StreamId) {
-        if let Some(ss) = self.0.get_mut(&stream_id) {
+        if let Some(ss) = self.map.get_mut(&stream_id) {
             ss.reset_lost();
         }
     }
 
     pub fn blocked_lost(&mut self, stream_id: StreamId, limit: u64) {
-        if let Some(ss) = self.0.get_mut(&stream_id) {
+        if let Some(ss) = self.map.get_mut(&stream_id) {
             ss.blocked_lost(limit);
         }
     }
 
     pub fn clear(&mut self) {
-        self.0.clear()
+        self.map.clear();
+        self.sendordered.clear();
+        self.regular.stream_ids().clear();
+    }
+
+    pub fn remove_terminal(&mut self) {
+        Self::remove_terminal_internal(&mut self.map, &mut self.regular, &mut self.sendordered);
+    }
+
+    fn remove_terminal_internal(map: &mut IndexMap<StreamId, SendStream>, regular: &mut OrderGroup, sendordered: &mut BTreeMap<SendOrder, OrderGroup>) {
```

Comment:
You can inline this function and save the complication of the function invocation.  You will still need to grab the mutable references outside of the retain loop.